File: backend/app/routers/market.py
from __future__ import annotations

from typing import Literal

from fastapi import APIRouter, Depends, Query
from pydantic import BaseModel, Field
from sqlalchemy.orm import Session

from app.database import get_db
from app.models.sector import SectorBasic, SectorDailyQuote, StockSectorMap
from app.exceptions import AppError
from app.services.main_wave_service import analyze_main_wave_stock, scan_main_wave_pool
from app.services.main_wave_sector_backfill_service import (
    get_main_wave_sector_backfill_status,
    run_main_wave_sector_backfill,
)
from app.services.sector_data_service import sync_sector_data
from app.tasks.background import get_task_status, submit_task
from app.services.trading_session import is_a_share_trading_session

router = APIRouter(prefix="/api/market", tags=["market"])


class SectorSyncRequest(BaseModel):
    sector_types: list[Literal["concept", "industry"]] = Field(default_factory=lambda: ["concept", "industry"])
    sync_constituents: bool = True
    sync_quotes: bool = True
    days: int = Field(180, ge=1, le=1200)
    limit: int | None = Field(30, ge=1, le=500)


class MainWaveSectorBackfillRequest(BaseModel):
    pool_id: str | None = None
    days: int = Field(250, ge=1, le=1200)
    mode: Literal["backfill", "incremental"] = "backfill"
    force: bool = False


class MainWaveBatchRequest(BaseModel):
    ts_codes: list[str] = Field(default_factory=list, max_length=100)


@router.get("/trading-session")
def get_trading_session():
    """与扫描服务同源的交易时段判定，供前端轻量轮询。"""
    return {"in_session": is_a_share_trading_session()}


@router.post("/sectors/sync")
def sync_sectors(body: SectorSyncRequest, db: Session = Depends(get_db)):
    """同步板块基础、成分和日线数据。limit 用于控制本次处理的板块数量。"""
    return sync_sector_data(
        db,
        sector_types=body.sector_types,
        sync_constituents=body.sync_constituents,
        sync_quotes=body.sync_quotes,
        days=body.days,
        limit=body.limit,
    )


@router.get("/sectors")
def list_sectors(
    sector_type: Literal["concept", "industry"] | None = Query(None),
    keyword: str | None = Query(None),
    limit: int = Query(100, ge=1, le=500),
    db: Session = Depends(get_db),
):
    q = db.query(SectorBasic)
    if sector_type:
        q = q.filter(SectorBasic.sector_type == sector_type)
    if keyword:
        q = q.filter(SectorBasic.sector_name.like(f"%{keyword}%"))
    rows = q.order_by(SectorBasic.rank.asc().nullslast(), SectorBasic.sector_name.asc()).limit(limit).all()
    return [
        {
            "sector_code": r.sector_code,
            "sector_name": r.sector_name,
            "sector_type": r.sector_type,
            "source": r.source,
            "raw_code": r.raw_code,
            "rank": r.rank,
            "latest_pct_chg": r.latest_pct_chg,
            "latest_hot": r.latest_hot,
            "updated_at": r.updated_at.isoformat() + "Z" if r.updated_at else None,
        }
        for r in rows
    ]


@router.get("/sectors/{sector_code}/quotes")
def list_sector_quotes(
    sector_code: str,
    limit: int = Query(120, ge=1, le=1200),
    db: Session = Depends(get_db),
):
    rows = (
        db.query(SectorDailyQuote)
        .filter(SectorDailyQuote.sector_code == sector_code)
        .order_by(SectorDailyQuote.trade_date.desc())
        .limit(limit)
        .all()
    )
    return [
        {
            "sector_code": r.sector_code,
            "trade_date": r.trade_date,
            "sector_name": r.sector_name,
            "sector_type": r.sector_type,
            "open": r.open,
            "high": r.high,
            "low": r.low,
            "close": r.close,
            "pct_chg": r.pct_chg,
            "change": r.change,
            "vol": r.vol,
            "amount": r.amount,
            "turnover_rate": r.turnover_rate,
        }
        for r in rows
    ]


@router.get("/stocks/{ts_code}/sectors")
def list_stock_sectors(
    ts_code: str,
    sector_type: Literal["concept", "industry"] | None = Query(None),
    db: Session = Depends(get_db),
):
    q = db.query(StockSectorMap).filter(StockSectorMap.ts_code == ts_code.upper())
    if sector_type:
        q = q.filter(StockSectorMap.sector_type == sector_type)
    rows = q.order_by(StockSectorMap.sector_type.asc(), StockSectorMap.sector_name.asc()).all()
    return [
        {
            "ts_code": r.ts_code,
            "sector_code": r.sector_code,
            "sector_name": r.sector_name,
            "sector_type": r.sector_type,
            "source": r.source,
            "weight": r.weight,
            "updated_at": r.updated_at.isoformat() + "Z" if r.updated_at else None,
        }
        for r in rows
    ]


@router.get("/stocks/{ts_code}/main-wave")
def analyze_stock_main_wave(ts_code: str, db: Session = Depends(get_db)):
    """单只股票主升浪趋势评分：个股趋势 + MA20 修复状态 + 板块共振。"""
    return analyze_main_wave_stock(db, ts_code)


@router.post("/main-wave/analyze-batch")
def analyze_main_wave_batch(body: MainWaveBatchRequest, db: Session = Depends(get_db)):
    """按前端当前列表批量评分，避免大池子打开时触发全池扫描。"""
    items = []
    seen: set[str] = set()
    for raw_code in body.ts_codes:
        ts_code = str(raw_code or "").strip().upper()
        if not ts_code or ts_code in seen:
            continue
        seen.add(ts_code)
        try:
            items.append(analyze_main_wave_stock(db, ts_code, allow_external_sector_fetch=False))
        except Exception:
            items.append(
                {
                    "ts_code": ts_code,
                    "name": ts_code,
                    "status": "insufficient_data",
                    "total_score": 0,
                    "message": "主升浪评分失败",
                }
            )
    return {"total": len(items), "items": items}


@router.get("/main-wave/scan")
def scan_main_wave(
    pool_id: str | None = Query(None),
    min_score: int | None = Query(None, ge=0, le=100),
    status: list[str] | None = Query(None),
    db: Session = Depends(get_db),
):
    """批量扫描主升浪评分，可按观察池、最低分和阶段过滤。"""
    return scan_main_wave_pool(db, pool_id=pool_id, min_score=min_score, statuses=status)


@router.get("/main-wave/sectors/backfill/status")
def get_main_wave_sector_status(
    pool_id: str | None = Query(None),
    days: int = Query(250, ge=1, le=1200),
    db: Session = Depends(get_db),
):
    """查看当前主升浪池相关概念板块 K 线补齐状态。"""
    return get_main_wave_sector_backfill_status(db, pool_id=pool_id, target_days=days)


@router.post("/main-wave/sectors/backfill")
def start_main_wave_sector_backfill(body: MainWaveSectorBackfillRequest):
    """后台补齐当前主升浪池相关概念板块 K 线，支持断点续跑和增量同步。"""
    task_id = submit_task(
        "main_wave_sector_backfill",
        run_main_wave_sector_backfill,
        pool_id=body.pool_id,
        days=body.days,
        mode=body.mode,
        force=body.force,
    )
    return {"task_id": task_id}


@router.get("/main-wave/sectors/backfill/tasks/{task_id}")
def get_main_wave_sector_backfill_task(task_id: str):
    status = get_task_status(task_id)
    if not status:
        raise AppError(code=1004, message="任务不存在", status_code=404)
    return {
        "task_id": status.id,
        "type": status.type,
        "status": status.status,
        "progress": status.progress,
        "message": status.message,
        "result": status.result,
        "created_at": status.created_at.isoformat() + "Z",
    }

File: frontend/src/api/market.ts
import client from './client'

export interface MainWaveAnalysis {
  ts_code: string
  name: string
  industry?: string | null
  trade_date?: string
  status: string
  total_score: number
  message?: string
  scores?: {
    trend: number
    structure: number
    pullback_repair: number
    sector_resonance: number
    market_relative?: number
  }
  ma20_state?: {
    state: string
    break_days: number
    distance_pct?: number | null
  }
  metrics?: {
    latest_close?: number | null
    return_20d?: number | null
    return_60d?: number | null
    breakout_date?: string
    breakout_price?: number
    above_short_ma_days_10?: number
    max_drawdown_10d?: number
    best_sector?: {
      sector_code: string
      sector_name: string
      sector_type: string
      sector_return_20d?: number | null
      stock_return_20d?: number | null
      relative_strength_20d?: number | null
      sync_ratio_20d?: number | null
      stock_drawdown_20d?: number | null
      sector_drawdown_20d?: number | null
    } | null
    sectors?: Array<{
      sector_code?: string | null
      sector_name: string
      sector_type?: string | null
      latest_hot?: number | null
      latest_pct_chg?: number | null
    }>
    sector_count?: number
    market_proxy?: {
      group?: string
      label?: string
      status?: string
      latest_pct_chg?: number | null
      latest_relative_pct_chg?: number | null
      relative_strength_20d?: number | null
    }
  }
  reasons?: {
    trend?: string[]
    structure?: string[]
    pullback_repair?: string[]
    sector_resonance?: string[]
    market_relative?: string[]
  }
}

export const analyzeMainWaveStock = (tsCode: string) =>
  client.get<MainWaveAnalysis>(`/market/stocks/${tsCode}/main-wave`)

export const analyzeMainWaveBatch = (tsCodes: string[]) =>
  client.post<{ total: number; items: MainWaveAnalysis[] }>('/market/main-wave/analyze-batch', {
    ts_codes: tsCodes,
  })

export const scanMainWave = (params?: {
  pool_id?: string
  min_score?: number
  status?: string[]
}) => client.get<{
  total: number
  items: MainWaveAnalysis[]
  summary: Record<string, number>
}>('/market/main-wave/scan', { params })

export const syncSectors = (data: {
  sector_types?: Array<'concept' | 'industry'>
  sync_constituents?: boolean
  sync_quotes?: boolean
  days?: number
  limit?: number | null
}) => client.post<{
  source: string
  sector_types: string[]
  sector_count: number
  constituent_count: number
  quote_count: number
  failed: string[]
}>('/market/sectors/sync', data)

export interface SectorBasicItem {
  sector_code: string
  sector_name: string
  sector_type: 'concept' | 'industry' | string
  source: string
  raw_code?: string | null
  rank?: number | null
  latest_pct_chg?: number | null
  latest_hot?: number | null
  updated_at?: string | null
}

export const listSectors = (params?: {
  sector_type?: 'concept' | 'industry'
  keyword?: string
  limit?: number
}) => client.get<SectorBasicItem[]>('/market/sectors', { params })

export interface MainWaveSectorBackfillItem {
  sector_code: string
  sector_name: string
  sector_type: string
  source: string
  status: string
  target_days?: number | null
  quote_count: number
  first_trade_date?: string | null
  last_trade_date?: string | null
  required_trade_date?: string | null
  freshness_status?: string | null
  freshness_lag_days?: number | null
  freshness_warning?: string | null
  is_fresh?: boolean
  attempts: number
  last_error?: string | null
  next_retry_at?: string | null
  last_success_at?: string | null
  stocks: string[]
}

export interface MainWaveSectorBackfillStatus {
  pool_id?: string | null
  pool_name?: string | null
  stock_count: number
  concept_count: number
  required_trade_date?: string | null
  completed_count: number
  partial_count: number
  stale_count?: number
  cooldown_count: number
  missing_count: number
  items: MainWaveSectorBackfillItem[]
}

export interface MainWaveSectorBackfillTask {
  task_id: string
  type: string
  status: string
  progress: number
  message: string
  result?: {
    quote_count?: number
    map_count?: number
    skipped?: number
    stale?: number
    local_proxy?: number
    required_trade_date?: string | null
    failed?: Array<{
      sector_code: string
      sector_name: string
      error: string
      next_retry_at?: string | null
    }>
  } | null
  created_at: string
}

export const getMainWaveSectorBackfillStatus = (params?: {
  pool_id?: string
  days?: number
}) => client.get<MainWaveSectorBackfillStatus>('/market/main-wave/sectors/backfill/status', { params })

export const startMainWaveSectorBackfill = (data: {
  pool_id?: string
  days?: number
  mode?: 'backfill' | 'incremental'
  force?: boolean
}) => client.post<{ task_id: string }>('/market/main-wave/sectors/backfill', data)

export const getMainWaveSectorBackfillTask = (taskId: string) =>
  client.get<MainWaveSectorBackfillTask>(`/market/main-wave/sectors/backfill/tasks/${taskId}`)
